File: lexer/lexer.go
package lexer

import (
	"monkey/token"
)

type Lexer struct {
	input        string
	position     int  //Current position in the input, also points to currentChar
	readPosition int  //Current reading position, after currentChar
	currentChar  byte //Current character under examination
}

func NewLexer(input string) *Lexer {
	lexer := &Lexer{
		input: input,
	}
	lexer.readChar()
	return lexer
}

func isLetter(character byte) bool {
	return 'a' <= character && character <= 'z' || 'A' <= character && character <= 'Z' || character == '_'
}

func isDigit(character byte) bool {
	return '0' <= character && character <= '9'
}

func (lexer *Lexer) readChar() {
	if lexer.readPosition >= len(lexer.input) {
		lexer.currentChar = 0
	} else {
		lexer.currentChar = lexer.input[lexer.readPosition]
	}
	lexer.position = lexer.readPosition
	lexer.readPosition += 1
}

func (lexer *Lexer) readIdentifier() string {
	currentPosition := lexer.position
	for isLetter(lexer.currentChar) {
		lexer.readChar()
	}
	return lexer.input[currentPosition:lexer.position]
}

func (lexer *Lexer) readNumber() string {
	currentPosition := lexer.position
	for isDigit(lexer.currentChar) {
		lexer.readChar()
	}
	return lexer.input[currentPosition:lexer.position]
}

func (lexer *Lexer) peekChar() byte {
	if lexer.position >= len(lexer.input) {
		return 0
	} else {
		return lexer.input[lexer.readPosition]
	}
}

func (lexer *Lexer) skipWhitespace() {
	for lexer.currentChar == ' ' || lexer.currentChar == '\t' || lexer.currentChar == '\n' || lexer.currentChar == '\r' {
		lexer.readChar()
	}
}

func newToken(tokenType token.TokenType, character byte) token.Token {
	return token.Token{
		Type:    tokenType,
		Literal: string(character),
	}
}

func (lexer *Lexer) NextToken() token.Token {
	var t token.Token
	lexer.skipWhitespace()
	switch lexer.currentChar {
	case '=':
		if lexer.peekChar() == '=' {
			char := lexer.currentChar
			lexer.readChar()
			t = token.Token{Type: token.EQUALS, Literal: string(char) + string(lexer.currentChar)}
		} else {
			t = newToken(token.ASSIGN, lexer.currentChar)
		}
	case '+':
		t = newToken(token.PLUS, lexer.currentChar)
	case '-':
		t = newToken(token.MINUS, lexer.currentChar)
	case ',':
		t = newToken(token.COMMA, lexer.currentChar)
	case ';':
		t = newToken(token.SEMICOLON, lexer.currentChar)
	case '(':
		t = newToken(token.LPAREN, lexer.currentChar)
	case ')':
		t = newToken(token.RPAREN, lexer.currentChar)
	case '{':
		t = newToken(token.LBRACE, lexer.currentChar)
	case '}':
		t = newToken(token.RBRACE, lexer.currentChar)
	case '/':
		t = newToken(token.SLASH, lexer.currentChar)
	case '!':
		if lexer.peekChar() == '=' {
			char := lexer.currentChar
			lexer.readChar()
			t = token.Token{Type: token.NOT_EQUALS, Literal: string(char) + string(lexer.currentChar)}
		} else {
			t = newToken(token.BANG, lexer.currentChar)
		}
	case '<':
		t = newToken(token.LESS_THAN, lexer.currentChar)
	case '>':
		t = newToken(token.GREATER_THAN, lexer.currentChar)
	case '*':
		t = newToken(token.ASTERISK, lexer.currentChar)
	case 0:
		t.Type = token.EOF
		t.Literal = ""
	default:
		if isLetter(lexer.currentChar) {
			identifier := lexer.readIdentifier()
			t.Type = token.LookupIdentifier(identifier)
			t.Literal = identifier
			return t
		} else if isDigit(lexer.currentChar) {
			digit := lexer.readNumber()
			t.Type = token.INT
			t.Literal = digit
			return t
		} else {
			t = newToken(token.ILLEGAL, lexer.currentChar)
		}
	}
	lexer.readChar()
	return t
}

File: src/lexer.c
#include "lexer.h"
#include "token.h"
#include "utils.h"
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

Lexer *new_lexer(char *input) {
  Lexer *l = malloc(sizeof(Lexer));
  if (l == NULL) {
    perror("malloc:");
    return NULL;
  }
  memset(l, 0, sizeof(Lexer));
  l->input = input;
  l->input_size = strlen(l->input);
  read_char(l);
  return l;
}

void read_char(Lexer *l) {
  if (l->read_position >= l->input_size) {
    l->current_character = 0;
  } else {
    l->current_character = l->input[l->read_position];
  }
  l->position = l->read_position;
  l->read_position++;
}

Token next_token(Lexer *l) {
  Token t;
  switch (l->current_character) {
  case '=':
    t = (Token){.Type = ASSIGN, .Literal = "="};
    break;
  case '+':
    t = (Token){.Type = PLUS, .Literal = "+"};
    break;
  case '(':
    t = (Token){.Type = LPAREN, .Literal = "("};
    break;
  case ')':
    t = (Token){.Type = RPAREN, .Literal = ")"};
    break;
  case '{':
    t = (Token){.Type = LBRACE, .Literal = "{"};
    break;
  case '}':
    t = (Token){.Type = RBRACE, .Literal = "}"};
    break;
  case ',':
    t = (Token){.Type = COMMA, .Literal = ","};
    break;
  case ';':
    t = (Token){.Type = SEMICOLON, .Literal = ";"};
    break;
  case '0':
    t = (Token){.Type = EOF, .Literal = ""};
    break;
  }
  read_char(l);
  return t;
}

char *read_identifier(Lexer *l) {
  int identifier_len = 0;
  int position = l->position;
  while (is_letter(l->current_character)) {
    read_char(l);
    identifier_len++;
  }
  char *identifier = malloc(identifier_len + 1);
  if (identifier == NULL) {
    return NULL;
  }
  for (int i = 0; i < identifier_len; i++, position++) {
    identifier[i] = l->input[position];
  }
  identifier[identifier_len] = '\0';
  return identifier;
}

void print_lexer(Lexer *l) {
  printf("Lexer input: %s\n", l->input);
  printf("Lexer input size: %ld\n", l->input_size);
  printf("Lexer position: %d\n", l->position);
  printf("Lexer read position: %d\n", l->read_position);
  printf("Lexer current character: '%c'\n", l->current_character);
}
